File: crates/txline/src/validation/v2.rs
//! V2 score-validation response and payload helpers.

use serde::{Deserialize, Serialize};

use super::legacy::{
    FixtureSummaryInput, ScoreStat, ScoresBatchSummary, timestamp_ms_to_epoch_day,
};
use super::proof::{Hash32, ProofNode};
use crate::{Result, TxlineError};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScoresStatValidationV2Response {
    pub ts: i64,
    #[serde(default)]
    pub stats_to_prove: Vec<ScoreStat>,
    pub event_stat_root: Hash32,
    pub summary: ScoresBatchSummary,
    #[serde(default)]
    pub stat_proofs: Vec<Vec<ProofNode>>,
    #[serde(default)]
    pub sub_tree_proof: Vec<ProofNode>,
    #[serde(default)]
    pub main_tree_proof: Vec<ProofNode>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ScoresStatValidationV2 {
    requested_stat_keys: Vec<u32>,
    response: ScoresStatValidationV2Response,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StatLeafInput {
    pub stat: ScoreStat,
    pub stat_proof: Vec<ProofNode>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StatValidationInput {
    pub ts: i64,
    pub fixture_summary: FixtureSummaryInput,
    pub fixture_proof: Vec<ProofNode>,
    pub main_tree_proof: Vec<ProofNode>,
    pub event_stat_root: [u8; 32],
    pub stats: Vec<StatLeafInput>,
}

impl ScoresStatValidationV2 {
    pub fn from_response(
        requested_stat_keys: Vec<u32>,
        response: ScoresStatValidationV2Response,
    ) -> Result<Self> {
        if requested_stat_keys.is_empty() {
            return Err(TxlineError::invalid_input(
                "V2 stat validation requires at least one stat key",
            ));
        }
        if response.stats_to_prove.len() != requested_stat_keys.len() {
            return Err(TxlineError::validation(format!(
                "statsToProve length {} does not match requested statKeys length {}",
                response.stats_to_prove.len(),
                requested_stat_keys.len()
            )));
        }
        for (idx, (stat, requested_key)) in response
            .stats_to_prove
            .iter()
            .zip(requested_stat_keys.iter())
            .enumerate()
        {
            if stat.key != *requested_key {
                return Err(TxlineError::validation(format!(
                    "statsToProve[{idx}].key {} does not match requested statKeys[{idx}] {}",
                    stat.key, requested_key
                )));
            }
        }
        if response.stat_proofs.len() != response.stats_to_prove.len() {
            return Err(TxlineError::validation(format!(
                "statProofs length {} does not match statsToProve length {}",
                response.stat_proofs.len(),
                response.stats_to_prove.len()
            )));
        }
        Ok(Self {
            requested_stat_keys,
            response,
        })
    }

    pub fn requested_stat_keys(&self) -> &[u32] {
        &self.requested_stat_keys
    }

    pub fn stats_to_prove(&self) -> &[ScoreStat] {
        &self.response.stats_to_prove
    }

    pub fn stat_proofs(&self) -> &[Vec<ProofNode>] {
        &self.response.stat_proofs
    }

    pub fn response(&self) -> &ScoresStatValidationV2Response {
        &self.response
    }

    /// Timestamp used by `validateStatV2` and the `daily_scores_roots` PDA.
    pub fn target_ts(&self) -> i64 {
        self.response.summary.update_stats.min_timestamp
    }

    pub fn epoch_day(&self) -> Result<u16> {
        timestamp_ms_to_epoch_day(self.target_ts())
    }

    pub fn to_validation_input(&self) -> StatValidationInput {
        let stats = self
            .response
            .stats_to_prove
            .iter()
            .cloned()
            .zip(self.response.stat_proofs.iter().cloned())
            .map(|(stat, stat_proof)| StatLeafInput { stat, stat_proof })
            .collect();

        StatValidationInput {
            ts: self.target_ts(),
            fixture_summary: FixtureSummaryInput {
                fixture_id: self.response.summary.fixture_id,
                update_count: self.response.summary.update_stats.update_count,
                min_timestamp: self.response.summary.update_stats.min_timestamp,
                max_timestamp: self.response.summary.update_stats.max_timestamp,
                events_sub_tree_root: self.response.summary.event_stats_sub_tree_root.to_bytes(),
            },
            fixture_proof: self.response.sub_tree_proof.clone(),
            main_tree_proof: self.response.main_tree_proof.clone(),
            event_stat_root: self.response.event_stat_root.to_bytes(),
            stats,
        }
    }

    pub fn leading_subset(&self, len: usize) -> Result<StatValidationInput> {
        if len == 0 || len > self.response.stats_to_prove.len() {
            return Err(TxlineError::validation(
                "V2 payload subset length must be within the proved stat count",
            ));
        }
        let mut input = self.to_validation_input();
        input.stats.truncate(len);
        Ok(input)
    }
}

File: typescript/src/validation/v2.ts
import { InvalidInputError, ValidationPayloadError } from "../errors.js";
import {
  fixtureSummaryInput,
  timestampMsToEpochDay,
  type FixtureSummaryInput,
  type ScoreStat,
  type ScoresBatchSummary,
} from "./legacy.js";
import { decodeHash32, type Hash32Like, type ProofNode } from "./proof.js";

export interface ScoresStatValidationV2Response {
  readonly ts: number;
  readonly statsToProve?: ScoreStat[];
  readonly eventStatRoot: Hash32Like;
  readonly summary: ScoresBatchSummary;
  readonly statProofs?: ProofNode[][];
  readonly subTreeProof?: ProofNode[];
  readonly mainTreeProof?: ProofNode[];
}

export interface StatLeafInput {
  readonly stat: ScoreStat;
  readonly statProof: readonly ProofNode[];
}

export interface StatValidationInput {
  readonly ts: number;
  readonly fixtureSummary: FixtureSummaryInput;
  readonly fixtureProof: readonly ProofNode[];
  readonly mainTreeProof: readonly ProofNode[];
  readonly eventStatRoot: Uint8Array;
  readonly stats: readonly StatLeafInput[];
}

export class ScoresStatValidationV2 {
  readonly #requestedStatKeys: readonly number[];
  readonly #response: ScoresStatValidationV2Response;
  readonly #statsToProve: readonly ScoreStat[];
  readonly #statProofs: readonly ProofNode[][];

  private constructor(
    requestedStatKeys: readonly number[],
    response: ScoresStatValidationV2Response,
    statsToProve: readonly ScoreStat[],
    statProofs: readonly ProofNode[][],
  ) {
    this.#requestedStatKeys = requestedStatKeys;
    this.#response = response;
    this.#statsToProve = statsToProve;
    this.#statProofs = statProofs;
  }

  static fromResponse(
    requestedStatKeys: readonly number[],
    response: ScoresStatValidationV2Response,
  ): ScoresStatValidationV2 {
    if (requestedStatKeys.length === 0) {
      throw new InvalidInputError(
        "V2 stat validation requires at least one stat key",
      );
    }
    const statsToProve = response.statsToProve ?? [];
    if (statsToProve.length !== requestedStatKeys.length) {
      throw new ValidationPayloadError(
        `statsToProve length ${statsToProve.length} does not match requested statKeys length ${requestedStatKeys.length}`,
      );
    }
    statsToProve.forEach((stat, index) => {
      const requested = requestedStatKeys[index];
      if (stat.key !== requested) {
        throw new ValidationPayloadError(
          `statsToProve[${index}].key ${stat.key} does not match requested statKeys[${index}] ${requested}`,
        );
      }
    });
    const statProofs = response.statProofs ?? [];
    if (statProofs.length !== statsToProve.length) {
      throw new ValidationPayloadError(
        `statProofs length ${statProofs.length} does not match statsToProve length ${statsToProve.length}`,
      );
    }
    return new ScoresStatValidationV2(
      [...requestedStatKeys],
      response,
      statsToProve,
      statProofs,
    );
  }

  requestedStatKeys(): readonly number[] {
    return this.#requestedStatKeys;
  }

  statsToProve(): readonly ScoreStat[] {
    return this.#statsToProve;
  }

  statProofs(): readonly ProofNode[][] {
    return this.#statProofs;
  }

  response(): ScoresStatValidationV2Response {
    return this.#response;
  }

  targetTs(): number {
    return this.#response.summary.updateStats.minTimestamp;
  }

  epochDay(): number {
    return timestampMsToEpochDay(this.targetTs());
  }

  toValidationInput(): StatValidationInput {
    const pseudoLegacy = {
      ts: this.#response.ts,
      statToProve: this.#statsToProve[0] ?? { key: 0, value: 0, period: 0 },
      eventStatRoot: this.#response.eventStatRoot,
      summary: this.#response.summary,
    };
    return {
      ts: this.targetTs(),
      fixtureSummary: fixtureSummaryInput(pseudoLegacy),
      fixtureProof: this.#response.subTreeProof ?? [],
      mainTreeProof: this.#response.mainTreeProof ?? [],
      eventStatRoot: decodeHash32(this.#response.eventStatRoot),
      stats: this.#statsToProve.map((stat, index) => ({
        stat,
        statProof: this.#statProofs[index] ?? [],
      })),
    };
  }

  leadingSubset(length: number): StatValidationInput {
    if (
      !Number.isInteger(length) ||
      length <= 0 ||
      length > this.#statsToProve.length
    ) {
      throw new ValidationPayloadError(
        "V2 payload subset length must be within the proved stat count",
      );
    }
    const input = this.toValidationInput();
    return {
      ...input,
      stats: input.stats.slice(0, length),
    };
  }
}
